### PR TITLE
Add a pjrt plugin device client to the XLA client.

### PR DESCRIPTION
**Pull request by jakeh-gc**
This pull request adds an opaque "plugin device client" to the XLA client interface.
This allows experimental/out-of-tree backends (perhaps a Graphcore IPU backend) to be exposed to users of the XLA client interface (like JAX) in a standard way.

This change is guarded by the `enable_plugin_device ` flag which defaults to `False`.

/cc @DavidNorman

### CODE REVIEW

File: tensorflow/compiler/xla/pjrt/pjrt_plugin_device_client.h
Comment (skye) [inline tensorflow/compiler/xla/pjrt/pjrt_plugin_device_client.h:27]: How/where is the plugin linked in? Is the idea you would build your own PJRT client binary that would also include the plugin client definition?

**Review by skye — COMMENTED** [review]
We're actually building a C version of the PJRT C API to better support different hardware/compiler/runtime plugins: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/pjrt/c/pjrt_c_api.h

This will include a C --> C++ connector, so you can still implement the PJRT C++ API if you like:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/pjrt/c/pjrt_c_api_wrapper_impl.h
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/pjrt/c/pjrt_c_api_wrapper_impl.cc

It's still WIP, and probably won't be ready for real use for another few months. If you can wait, I'd prefer to not merge this C++ mechanism, which I think will still require custom jaxlib or other client binaries (see my comment). However, I can be convinced to merge this and revert it once the C-based plugin mechanism is ready.

File: tensorflow/compiler/xla/pjrt/pjrt_plugin_device_client.h
Comment (jakeh-gc) [reply tensorflow/compiler/xla/pjrt/pjrt_plugin_device_client.h:27]: Yes, correct. It would be linked through `//tensorflow/compiler/plugin:plugin`, which is reserved for third-party XLA backends. This is exactly where the IPU XLA backend lives in our fork. Here's an example of a minimal dummy implementation to make it compile and link https://github.com/jakeh-gc/tensorflow/commit/c9772a1017d8b83aeb5e72044578336e31c49def

**Comment by jakeh-gc** [comment]
> We're actually building a C version of the PJRT C API to better support different hardware/compiler/runtime plugins: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/pjrt/c/pjrt_c_api.h
> 
> This will include a C --> C++ connector, so you can still implement the PJRT C++ API if you like: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/pjrt/c/pjrt_c_api_wrapper_impl.h https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/pjrt/c/pjrt_c_api_wrapper_impl.cc

I think that looks good, and we would certainly like a stable ABI boundary. The thing that isn't clear to me from the linked C API is how a backend is registered and made available to users.

The main goal of this pull request is to make third-party pjrt clients available to users of the xla_client similar to how CPUs, GPUs, and TPUs are.

> I think will still require custom jaxlib or other client binaries
This is true, the corresponding pull request for jax is here https://github.com/google/jax/pull/11961

With that, I can locally build with `python build/build.py --enable_plugin_device` and then use IPUs in jax without any upstream changes (besides that PR of course). Currently this requires maintaining a separate fork that doesn't always merge cleanly.

That being said, I'm happy to defer to your judgement if there's a better way.

**Comment by skye** [comment]
> The thing that isn't clear to me from the linked C API is how a backend is registered and made available to users.

We haven't designed let alone implemented that yet :) Suggestions welcome if you have ideas or existing plugin APIs you like! I'm vaguely imagining a low-level mechanism for specifying where to look for plugin libraries (e.g. env var and/or default paths), that can ideally be used to build user-friendlier plugin installation (e.g. pip).

> Currently this requires maintaining a separate fork that doesn't always merge cleanly. That being said, I'm happy to defer to your judgement if there's a better way.

I'm ok merging this to make your life easier while we implement the C plugin interface. I may not have time to thoroughly review until next week though (I'm OOO tomorrow).

**Comment by jakeh-gc** [comment]
> I'm ok merging this to make your life easier while we implement the C plugin interface.

Okay, that sounds reasonable. I'm happy for this to be merged with the understanding it will be removed when the C plugin interface is available.

Same for the corresponding PR in jax https://github.com/google/jax/pull/11961

File: tensorflow/compiler/xla/python/xla_client.py
Comment (skye) [inline tensorflow/compiler/xla/python/xla_client.py:116]: Maybe catch if this isn't defined and mention the build flag?

File: tensorflow/compiler/xla/python/xla.cc
Comment (skye) [inline tensorflow/compiler/xla/python/xla.cc:349]: nit: this function is quite short, remove blank lines (here and below)

File: tensorflow/compiler/xla/pjrt/pjrt_plugin_device_client.h
Comment (skye) [reply tensorflow/compiler/xla/pjrt/pjrt_plugin_device_client.h:27]: > It would be linked through //tensorflow/compiler/plugin:plugin, which is reserved for third-party XLA backends.

Can you mention this in the comment as well?

**Review by skye — CHANGES_REQUESTED** [review]
Looks good! A few minor comments

**Review by skye — APPROVED** [review]

**Review by skye — APPROVED** [review]

**Comment by jakeh-gc** [comment]
Any idea what I can do to resolve this?
> feedback/copybara — Google internal checks FAILED for runs with create time 2022-09-06T16:34:12.870440305Z.

I also need to resolve a merge conflict.

**Comment by skye** [comment]
Hi Jake, sorry for the delay on this. I think the internal failure is due to the merge conflict. I think I can actually resolve it from my side with Google magic, lemme try that.

**Comment by skye** [comment]
Still working on this. There are some unrelated Google-internal failures that are blocking getting this in. You don't need to do anything (except maybe ping me in a few days if this doesn't get go in :)).

File: tensorflow/compiler/xla/pjrt/BUILD
Comment (joker-eph) [inline tensorflow/compiler/xla/pjrt/BUILD:457]: I need to break dependencies from XLA to TensorFlow, and so this one needs to go away. What is the best direction for this: move `tensorflow/compiler/plugin` to XLA or move `pjrt_plugin_device_client` to `tensorflow/compiler/plugin`?

Comment (skye) [reply tensorflow/compiler/xla/pjrt/BUILD:457]: @qqfish @penpornk WDYT? I _think_ `tensorflow/compiler/plugin` would ideally move to XLA, but it might not be possible to move [this `plugin_device` rule](https://github.com/tensorflow/tensorflow/blob/a1478995fff6fc5d83932f90533f710a500d1c8d/tensorflow/compiler/plugin/BUILD#L45-L52), which might break actual plugins? Maybe we should just do it though.

`pjrt_plugin_device_client` is also a temporary solution while we get the PJRT C API plugin interface working, so we should probably just do whatever's easiest for now. Which I think would be moving `tensorflow/compiler/plugin` to XLA, since `pjrt_plugin_device_client` will have lots of PJRT deps.

Comment (jakeh-gc) [reply tensorflow/compiler/xla/pjrt/BUILD:457]: There's a real chance that I might be the only user of this interface. In the background I've been (slowly) catching up with removing TF dependencies (in sync with upstream) from my plugin device. I think I could survive the move to XLA.

Is there anything I can look at to see whether what you propose would be breaking?

Comment (jakeh-gc) [reply tensorflow/compiler/xla/pjrt/BUILD:457]: For some added context @joker-eph, the only reason I chose that directory for this rule was because that's the canonical place for third-party backends to live in the TF repo. That being different in XLA makes sense.

Comment (joker-eph) [reply tensorflow/compiler/xla/pjrt/BUILD:457]: Since I don't have much knowledge of this all, my easy "fix" would be to just drop this dependency (`"//tensorflow/compiler/plugin",`), but you may want instead to move code one way or another?

If you have an idea, please send a pull-request?

Comment (jakeh-gc) [reply tensorflow/compiler/xla/pjrt/BUILD:457]: Sure. I can put up a PR in the morning.
Just to be clear, you want no dependencies outside the `tensorflow/compiler/xla` directory or do you want this somewhere else entirely?

Comment (jakeh-gc) [reply tensorflow/compiler/xla/pjrt/BUILD:457]: This is the kind of change I mean. https://github.com/jakeh-gc/tensorflow/commit/1de7e76f685fb4c378b7f660ff7ae81bc0ae8200
I'm in the U.K. and out-of-office until the morning, so I haven't checked it builds yet.

Comment (joker-eph) [reply tensorflow/compiler/xla/pjrt/BUILD:457]: Looks good! Thanks :)

Comment (penpornk) [reply tensorflow/compiler/xla/pjrt/BUILD:457]: Re: @skye:

> @qqfish @penpornk WDYT?

+1 on moving `tensorflow/compiler/plugin` to XLA like in https://github.com/jakeh-gc/tensorflow/commit/1de7e76f685fb4c378b7f660ff7ae81bc0ae8200. I'm not familiar with how `tensorflow/compiler/plugin` is being used, but if @jakeh-gc is the only user, we should be fine.

Comment (jakeh-gc) [reply tensorflow/compiler/xla/pjrt/BUILD:457]: Created a PR for this #58600

**Pull request merged**
